Cloud transcription

WingCH opened the pull request:
https://github.com/Beingpax/VoiceInk/issues/11

This pull request implements several key improvements to the VoiceInk codebase including standardized UserDefaults access, cloud transcription integration.

# Key Changes
## Cloud Transcription Service
* Added CloudTranscriptionService implementation using OpenAI's GPT-4o-transcribe model
* Created TranscriptionService protocol for unified interface across transcription services
* Implemented TranscriptionServiceFactory for managing service instantiation
* Added UI controls for selecting between local and cloud transcription services

## UserDefaults Standardization
* Replaced direct UserDefaults access with structured properties across all files
* Created UserDefaultsKeys enum for consistent key naming
* Added typed properties for safer and more maintainable UserDefaults access

Beingpax commented:
![CleanShot 2025-03-29 at 21 06 43@2x](https://github.com/user-attachments/assets/bb0153ef-4088-4f2b-9682-57abcf06ba6a)

I think we need to fix this error. When the language is set as auto-detect (because OpenAI does not have this parameter), I think we need to remove the language parameter itself for OpenAI if we want to use auto-detect.

WingCH commented:
The issue with the 'auto' language code has been fixed.

I also addressed a couple of other things:

1. The animation now properly reflects the isProcessing state.

before

<img width="207" alt="image" src="https://github.com/user-attachments/assets/5842645e-d1c0-4b64-b02a-2d18225ef3c1" />


after
<img width="183" alt="image" src="https://github.com/user-attachments/assets/46956b00-7839-43c4-b848-1b930424cade" />


2. On MacBooks, one of the shortcut keys for switching input languages is fn, which conflicted with our default setting for "Push-to-Talk". I’ve updated it so that it only triggers when fn is held for more than 0.25 seconds—this should prevent accidental activation while switching languages.

Beingpax commented:
![CleanShot 2025-03-30 at 10 32 09@2x](https://github.com/user-attachments/assets/60efaa5e-5fe1-4221-9164-ec15a1804780)
Sometimes the processing goes on forever. There's this issue in the debug area.

> _quic_packet_builder_calculate_size packet builder has frames, but none of those frames fit in this packet


Also I think we need to add a simple retry mechanism for AI requests in case of failed requests. 

![Xcode 2025-03-30 10 35 54](https://github.com/user-attachments/assets/4aed10fc-5f25-431e-93f1-72cfa7f9ab62)

WingCH commented:
I just added retry mechanism

For the error message "_quic_packet_builder_calculate_size: packet builder has frames, but none of those frames fit in this packet," I am unable to reproduce it. Could you please share the steps to reproduce this error?

abotmaker commented on VoiceInk/Views/MenuBarView.swift:
```diff
@@ -54,39 +54,99 @@ struct MenuBarView: View {
+                }
+            }
+            
+            if whisperState.transcriptionServiceType == .cloud {
```
is it possible to include the gpt-4o-mini-transcribe
too? this model offers a slight trade-off in accuracy for half the price wrt gpt-4o-transcribe(https://platform.openai.com/docs/models/gpt-4o-mini-transcribe). so could be an interesting add. thank you!

abotmaker commented on VoiceInk/Views/MenuBarView.swift:
```diff
@@ -54,39 +54,99 @@ struct MenuBarView: View {
+            
+            if whisperState.transcriptionServiceType == .cloud {
+                Menu {
+                    let availableCloudModels = ["gpt-4o-transcribe", "whisper-1"]
```
is it possible to include the gpt-4o-mini-transcribe
too? this model offers a slight trade-off in accuracy for half the price wrt gpt-4o-transcribe(https://platform.openai.com/docs/models/gpt-4o-mini-transcribe). so could be an interesting add. thank you!

WingCH replied:
sure, just added

Beingpax replied:
Here's a quick Video. The issue is random. the recent changes has solved this so it doesn't get stuck in forever. But the issue:

> _quic_packet_builder_calculate_size packet builder has frames, but none of those frames fit in this packet

still happens. And takes a significant time to do the next retry. Also this happens in cases when the full audio transcription fails. So I don't if this issue is related to our audio file. 


https://youtu.be/MQvuJg1oN0o


🔄 Starting transcription with service type: Cloud
💬 Prompt set: 
Important words: VoiceInk
🌐 Language set: auto
🔄 Starting OpenAI transcription from file URL using gpt-4o-transcribe
_quic_packet_builder_calculate_size packet builder has frames, but none of those frames fit in this packet
_quic_packet_builder_calculate_size packet builder has frames, but none of those frames fit in this packet
Handled error from server: Connection interrupted
quic_conn_keepalive_handler [C33.1.1.1:2] [-01b22b20e00de4f5bcb2a520ca0dca0046b5810c] keep-alive timer fired, exceeding 2 outstanding keep-alives
nw_read_request_report [C33] Receive failed with error "Operation timed out"
nw_read_request_report [C33] Receive failed with error "Operation timed out"
nw_read_request_report [C33] Receive failed with error "Operation timed out"
nw_read_request_report [C33] Receive failed with error "Operation timed out"
Connection 33: write error 1:89
Connection 33: received failure notification
Task <4AF67DFC-7EE9-4AB0-9D0F-06E91DB3D7A8>.<70> HTTP load failed, 520/0 bytes (error code: -1001 [1:60])
Task <4AF67DFC-7EE9-4AB0-9D0F-06E91DB3D7A8>.<70> finished with error [-1001] Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={_kCFStreamErrorCodeKey=60, NSUnderlyingError=0x60000117b600 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={NSErrorPeerAddressKey=<CFData 0x600003c6fed0 [0x20a446240]>{length = 16, capacity = 16, bytes = 0x100201bba29f8cf50000000000000000}, _kCFStreamErrorCodeKey=60, _kCFStreamErrorDomainKey=1}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <4AF67DFC-7EE9-4AB0-9D0F-06E91DB3D7A8>.<70>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <4AF67DFC-7EE9-4AB0-9D0F-06E91DB3D7A8>.<70>"
), NSLocalizedDescription=The request timed out., NSErrorFailingURLStringKey=https://api.openai.com/v1/audio/transcriptions, NSErrorFailingURLKey=https://api.openai.com/v1/audio/transcriptions, _kCFStreamErrorDomainKey=1}
Unknown error type encountered: The request timed out.
⏱️ Retrying transcription (attempt 1/3) after 1.00 seconds. Error: Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={_kCFStreamErrorCodeKey=60, NSUnderlyingError=0x60000117b600 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={NSErrorPeerAddressKey=<CFData 0x600003c6fed0 [0x20a446240]>{length = 16, capacity = 16, bytes = 0x100201bba29f8cf50000000000000000}, _kCFStreamErrorCodeKey=60, _kCFStreamErrorDomainKey=1}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <4AF67DFC-7EE9-4AB0-9D0F-06E91DB3D7A8>.<70>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <4AF67DFC-7EE9-4AB0-9D0F-06E91DB3D7A8>.<70>"
), NSLocalizedDescription=The request timed out., NSErrorFailingURLStringKey=https://api.openai.com/v1/audio/transcriptions, NSErrorFailingURLKey=https://api.openai.com/v1/audio/transcriptions, _kCFStreamErrorDomainKey=1}
nw_endpoint_handler_unregister_context [C33.1.1.1 162.159.140.245:443 failed channel-flow (satisfied (Path is satisfied), viable, interface: en0[802.11], ipv4, ipv6, dns, uses wifi)] Cannot unregister after flow table is released
nw_endpoint_flow_fillout_data_transfer_snapshot copy_info() returned NULL
✅ OpenAI transcription successfully completed
✅ Transcription completed successfully, text: ७, ६, ५., length: 8 characters
✅ Word replacements applied
📱 Dismissing mini recorder
Command+V paste completed
Stopping recording
Triggering audio device change notification
Recording stopped successfully
Handling device change
Device change handling completed
🧹 Cleaning up Whisper resources
🧹 OpenAI transcription resources released
Stopping recording
Triggering audio device change notification
Recording stopped successfully
Handling device change
Device change handling completed
🧹 Cleaning up Whisper resources

Beingpax replied:
_quic_packet_builder_calculate_size packet builder has frames, but none of those frames fit in this packet

This issue happened with me as well when I was trying to implement it a few weeks back. But couldn't get past it. So I didn't spend more time on it because there was not significant improvement in transcription quality.

WingCH replied:
I tried using the History feature to regenerate the transcription but couldn't reproduce the issue. I never see the `_quic_packet_builder_calculate_size` error message. 😵

Using the cloud model for my languages, such as Chinese or Cantonese, has shown a very significant improvement. You might be using English, so the effect has always been good. 😭

Can you share some information?
1. your MacBook model
2. macOS version
3. analyze network calls using Proxyman or similar network inspection tools to look for potential hints or clues.
4. your country to check if network issues are causing the API to frequently fail
5. openAI account tier

Network call, example
<img width="1548" alt="Proxyman 2025-03-30 22 51 54" src="https://github.com/user-attachments/assets/220faaf9-e784-457d-9d1c-21c6e56c23db" />

Beingpax commented:
Version 16.2 (16C5032a): Xcode
15.3.2 (24D81): Macbook Pro M2

It is not happening anymore. And I can't figure out any particular reason this might be happening. But there's this another issue. 

Sometimes it works properly, but other times, it only gives me prompt text back instead of transcibing the actual audio file.


https://github.com/user-attachments/assets/03452eac-a4ee-478c-9334-16e542c72ef2

WingCH commented:
Haha, I've also encountered this issue. 

I found that when using the API,  
there are two situations where this problem is fully discovered:
1. The sound is unclear.  
2. The audio is very short and is for testing purposes, such as "test, test."  

If you remove the prompt, it can solve this problem. You can try selecting the automatic language, as the automatic language doesn't have prompts.  

Do you want me to use the API without adding a prompt?

Beingpax commented:
Yes but we still should pass the dictionary words. Will that also impact it?

On Mon, 31 Mar 2025 at 03:49 WingCH ***@***.***> wrote:

> Haha, I've also encountered this issue.
>
> I found that when using the API,
> there are two situations where this problem is fully discovered:
>
>    1. The sound is unclear.
>    2. The audio is very short and is for testing purposes, such as "test,
>    test."
>
> If you remove the prompt, it can solve this problem. You can try selecting
> the automatic language, as the automatic language doesn't have prompts.
>
> Do you want me to use the API without adding a prompt?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Beingpax/VoiceInk/pull/42#issuecomment-2764766216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYCUXQBIHLIYZ5YSSDUCGW32XBS6TAVCNFSM6AAAAAB2BMJLI6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUG43DMMRRGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: WingCH]*WingCH* left a comment (Beingpax/VoiceInk#42)
> <https://github.com/Beingpax/VoiceInk/pull/42#issuecomment-2764766216>
>
> Haha, I've also encountered this issue.
>
> I found that when using the API,
> there are two situations where this problem is fully discovered:
>
>    1. The sound is unclear.
>    2. The audio is very short and is for testing purposes, such as "test,
>    test."
>
> If you remove the prompt, it can solve this problem. You can try selecting
> the automatic language, as the automatic language doesn't have prompts.
>
> Do you want me to use the API without adding a prompt?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Beingpax/VoiceInk/pull/42#issuecomment-2764766216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYCUXQBIHLIYZ5YSSDUCGW32XBS6TAVCNFSM6AAAAAB2BMJLI6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUG43DMMRRGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Beingpax commented:
Hey, this is not something completely related to this issue, but I have been trying wispr flow for the past few days. One thing that I really like about it is that it is able to distinguish multiple languages.

I guess the Whisper model by default does not allow you to output in 2 languages. So I guess they are using a streaming service and using auto detect as a language so that it will automatically detect when the language is changed.

How do you think they are able to accomplish this if you have any idea? I really want to work on this because this will be much better when using it for multilingual languages because you will get a output in multiple languages, instead of just getting it in one language output.

WingCH commented:
I will remove the prompt tonight.

You mean Wispr Flow can output more than one language?
I remember their settings require setting a preferred language, so it should only output one language?

Beingpax commented:
I am moving out right now. So I've packed my mac already else I would send
you screenshot. Lets say you add english and hindi as putput language(I
have tried that) It will properly detect when I'm talking in hindi or
english. Try it out or else I will share my history in the next 2 days

On Mon, 31 Mar 2025 at 11:18 WingCH ***@***.***> wrote:

> I will remove the prompt tonight.
>
> You mean Wispr Flow can output more than one language?
> I remember their settings require setting a preferred language, so it
> should only output one language?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Beingpax/VoiceInk/pull/42#issuecomment-2765191323>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYCUXQDAHBLY74FM54CB3QD2XDHSJAVCNFSM6AAAAAB2BMJLI6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVGE4TCMZSGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: WingCH]*WingCH* left a comment (Beingpax/VoiceInk#42)
> <https://github.com/Beingpax/VoiceInk/pull/42#issuecomment-2765191323>
>
> I will remove the prompt tonight.
>
> You mean Wispr Flow can output more than one language?
> I remember their settings require setting a preferred language, so it
> should only output one language?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Beingpax/VoiceInk/pull/42#issuecomment-2765191323>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYCUXQDAHBLY74FM54CB3QD2XDHSJAVCNFSM6AAAAAB2BMJLI6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVGE4TCMZSGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

WingCH commented:
You mean to output in English when you speak English,
and output in Hindi when you speak Hindi?

Beingpax commented:
https://github.com/user-attachments/assets/ab5314f9-6053-4d8b-98b6-ad4732ffba9c

This is what I meant.

WingCH commented:
I think they are not using Whisper directly 😂 Maybe they are processing a piece of audio separately? It probably isn't easy.

Beingpax commented:
ok, Thanks for that info. I've reached home. And I'll be testing out and merge if everything looks fine. Thank you for your effort. And you were right, the cloud model has significant improvements for languages other than English. English was already good. 

P.S. I tested it with English, Hindi, and Nepali languages.

Beingpax commented:
I'm facing these errors. Have you also gone through it. The complete audio data is not being processed. @WingCH 

The full audio is not being processed. The left one is a simple web version and the right one is our app VoiceInk 

https://github.com/user-attachments/assets/9d66f5d0-626d-448f-ad55-f104a6e1a160

Beingpax commented:
This issue only happens with gpt-4o transcribe model. The same audio file when sent through whisper-1, it does the transcription perfectly. 

I just got superwhisper pro as well to check it out, and this issue happens there as well. But its less frequent. With VoiceInk it happens always. 

Also, found this on community forum, I will put this update on hold for now. 

 https://community.openai.com/t/gpt-4o-transcribe-truncates-the-transcript/1148347

WingCH commented:
I think we could list the limitations of the different models and then let the user choose?

Also, regarding Superwhisper Pro, I have tried it, but I couldn't find any place to use the OpenAI gpt-4o transcribe model.


<img width="583" alt="superwhisper 2025-04-06 09 50 55" src="https://github.com/user-attachments/assets/00c16782-2585-43c1-8646-dd4c7cca19b3" />
<img width="802" alt="image" src="https://github.com/user-attachments/assets/11bc3276-9b0c-490f-ac60-cd957b480e40" />

Beingpax commented:
https://github.com/user-attachments/assets/c932ac99-eb9e-4b2f-97d4-4935d6bd4ded

WingCH commented:
I see paid users can have this OpenAI option. I've already paid and am using it now, let's see what differences there are.

Regarding your issue, I believe the software Superwhisper has made some optimizations.

abotmaker commented:
from the exchange, it is clear that the issue is related to the OpenAI transcribe API. so in any case, I think we can go ahead with the release of this feature. In addition, we can also add the possibility to integrate the cloud transcription via API from other services like Eleven Labs and Groq Whisper as it is being done in this app ( https://github.com/braden-w/whispering ). 

Personally, I think this should be our priority because the cloud transcription model really optimizes for the RAM and accuracy.
